c2rust-analyze/src/rewrite/expr/hir_op.rs

```diff
@@ -326,29 +361,47 @@ fn mutbl_from_bool(m: bool) -> hir::Mutability {
     }
 }
 
-fn materialize_adjustment<'tcx>(
+fn apply_identity_adjustment<'tcx>(
     tcx: TyCtxt<'tcx>,
-    adj: &Adjustment<'tcx>,
-    hir_rw: Rewrite,
+    adjustment: &Adjustment<'tcx>,
+    rw: Rewrite,
 ) -> Rewrite {
-    match adj.kind {
-        Adjust::NeverToAny => {
-            // Should work fine with no explicit cast.
-            hir_rw
-        }
-        Adjust::Deref(_) => Rewrite::Deref(Box::new(hir_rw)),
-        Adjust::Borrow(AutoBorrow::Ref(_, mutbl)) => Rewrite::Ref(Box::new(hir_rw), mutbl.into()),
-        Adjust::Borrow(AutoBorrow::RawPtr(mutbl)) => Rewrite::AddrOf(Box::new(hir_rw), mutbl),
+    match adjustment.kind {
+        Adjust::NeverToAny => rw,
+        Adjust::Deref(_) => Rewrite::Deref(Box::new(rw)),
+        Adjust::Borrow(AutoBorrow::Ref(_, mutbl)) => Rewrite::Ref(Box::new(rw), mutbl.into()),
+        Adjust::Borrow(AutoBorrow::RawPtr(mutbl)) => Rewrite::AddrOf(Box::new(rw), mutbl),
         Adjust::Pointer(PointerCast::Unsize) => {
-            let ty = adj.target;
+            let ty = adjustment.target;
             let printer = FmtPrinter::new(tcx, Namespace::TypeNS);
             let s = ty.print(printer).unwrap().into_buffer();
-            Rewrite::Cast(Box::new(hir_rw), s)
+            Rewrite::Cast(Box::new(rw), s)
         }
         Adjust::Pointer(cast) => todo!("Adjust::Pointer({:?})", cast),
     }
 }
 
+fn materialize_adjustments<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    adjustments: &[Adjustment<'tcx>],
+    hir_rw: Rewrite,
+) -> Rewrite {
+    let adj_kinds: Vec<&_> = adjustments.iter().map(|a| &a.kind).collect();
+    match (hir_rw, &adj_kinds[..]) {
+        (Rewrite::Identity, []) => Rewrite::Identity,
+        (hir_rw @ Rewrite::Identity, _) => {
```

Comment:
This rebinding is really confusing--you already have a name for `hir_rw`, so the @-pattern isn't necessary here. Right?

Reply:
I've simplified it in 1608de9d. The shadowing is still necessary without the bind because `hir_rw` is moved in the match.